Feature/user spec changes

wcjr opened the pull request:
wanted to go ahead and get this functionality in place, tests still still to come.

of note: 

1.  pluralizes user endpoint, e.g., `GET /user` is now `GET /users`
2.  moves to traditional `id` usage instead of `employee_id`, e.g., `PUT /user/{employee_id}` is now `PUT /user/{id}`.  With the additional of the `GET /users` in this spec change, there's always a way to get the generated key for a given resource now.

resolves #31 core-functionality

forevermatt commented on api.raml:
```diff
@@ -55,6 +55,10 @@ schemas:
       {
         "$schema": "http://json-schema.org/draft-03/schema",
         "properties": {
+            "id": {
```
I think we actually will want to continue hiding our internal `id` primary key (since it's only used for things like foreign key relations). Since there are multiple different systems, many of them with databases, the `employee_id` is how we are uniquely and consistently identifying users across systems.

fillup replied:
agree

wcjr replied:
ok, I'll change that back.

forevermatt commented on api.raml:
```diff
@@ -231,32 +240,46 @@ schemas:
               }
 
-/user:
+/users:
```
For the sake of consistency, I think we typically keep the endpoint singular, knowing that a `GET` request without an individual record identifier is a request for a list of all of them.

fillup replied:
agree

wcjr replied:
ok, I'll change that back.

lvail commented:
Just to clarify for me: The endpoint PUT /user/{id} should still use
employee_id even though it is now called {id}?

On Mon, Apr 3, 2017 at 1:29 PM, forevermatt <notifications@github.com>
wrote:

> *@forevermatt* commented on this pull request.
> ------------------------------
>
> In api.raml
> <https://github.com/silinternational/idp-id-broker/pull/41#discussion_r109476531>
> :
>
> > @@ -55,6 +55,10 @@ schemas:
>        {
>          "$schema": "http://json-schema.org/draft-03/schema",
>          "properties": {
> +            "id": {
>
> I think we actually will want to continue hiding our internal id primary
> key (since it's only used for things like foreign key relations). Since
> there are multiple different systems, many of them with databases, the
> employee_id is how we are uniquely and consistently identifying users
> across systems.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/silinternational/idp-id-broker/pull/41#pullrequestreview-30581307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACA6243HCRspK1F3_OTPzcTeqFEduwGzks5rsSyGgaJpZM4Mx13n>
> .
>



-- 
Larry Vail
JAARS Volunteer, DevOps Engineer
JAARS Inc. Waxhaw, North Carolina

fillup commented:
@lvail that is correct, although @wcjr will need to confirm that the source works that way.

forevermatt commented on application/frontend/controllers/UserController.php:
```diff
@@ -8,11 +8,16 @@
 
+    public function actionView($employeeId) // GET /user/abc123
+    {
+        return User::findOne(['employee_id' => $employeeId]);
```
You mentioned that you might need to add more logic in order to return a `204` when no such user is found ([spec](https://github.com/silinternational/idp-id-broker/blob/develop/api.raml#L323)).

wcjr replied:
ok, I'll add that.

forevermatt commented on api.raml:
```diff
@@ -384,8 +407,20 @@ schemas:
-              The password for the user with that `employee_id` was
-              successfully updated.
+              The password for the user with that `id` was successfully updated.
+          404:
```
This might be a `204` to match our use of 404 vs. 204 elsewhere in the spec, but thanks for highlighting this. I hadn't thought about what happens when we try to update the password of a user that isn't found in the database.

wcjr replied:
let me know if you want a change here.

wcjr commented:
@lvail You weren't reading that `{id}` thing wrong.  I was under the assumption we were constrained in passing `employee_id` because that's the only thing id-sync would have.  When we added the "get all users" call and changed our syncing strategy I thought we could fallback to a more traditional REST convention.  Sounds like that change is going to be too disruptive at this point so I'll be changing the API back to passing around `employee_id` tomorrow.

wcjr commented:
@fillup @forevermatt here's the core functionality for splitting the API's.  Tests to follow later.

forevermatt commented on api.raml:
```diff
@@ -352,6 +361,22 @@ schemas:
+                  "display_name": "John Smith",
+                  "username": "john_smith",
+                  "email": "john_smith@example.com",
+                  "active": "yes",
```
@wcjr Since the example call to `PUT` has `"active": "no"`, what would you think of making `active` equal to `"no"` here in the response example? By no means necessary, it just seems like a nice nuance.

wcjr replied:
yep, that's better:  https://github.com/silinternational/idp-id-broker/commit/4eb6316744ab7da8da0c32f06af6ec23b3a554b1

forevermatt commented on api.raml:
```diff
@@ -320,8 +330,7 @@ schemas:
       responses:
         200:
           description:
```
I'm not positive, but I think this `description:` will need to be `description: |` since the text is on the next line.

wcjr replied:
it's in, https://github.com/silinternational/idp-id-broker/commit/fc791e5a55b97d62a70d1178b079d83b346fcee8.  Did your plugin flag that as a syntax error?  Mine didn't...

forevermatt replied:
No, mine didn't either. I have hit errors (I don't remember whether it was in my IDE or elsewhere) for things similar to this, though... that's all.

forevermatt commented on application/frontend/config/main.php:
```diff
@@ -30,14 +33,18 @@
                     'class' => 'yii\rest\UrlRule',
-                    'controller' => ['user', 'authentication'],
+                    'controller' => ['authentication', 'user'],
+                    'extraPatterns' => [
```
Will this `extraPatterns` entry apply to both the `authentication` and `user` controllers? I'm not sure I understand the way this is supposed to be used (specifically the patterns in this list, since I would have expected them to start with `/user` or `/authentication`).

wcjr replied:
while both controllers are in the mix and getting a bunch of standard patterns automatically generated for each of them from `UrlRule`, the `AuthenticationController` doesn't implement a `view` so even if an "authorized" requester were to issue a `GET /authentication/abc123`, the response will still be a `400` because of `BaseRestController`, specifically in this case, it will fail because there was no `id` passed along.

forevermatt replied:
I would have simply listed all half-dozen of the endpoints specifically, like the example on (old) line 38. That may just be leftover practices from my Yii1 tendencies, though. Feel free to go with whatever you prefer here.

wcjr replied:
not needing those `405`'s simplifies things greatly.  https://github.com/silinternational/idp-id-broker/commit/af1e2cd61eb2acb345276fdd9fad4552c87f7524

forevermatt commented on application/frontend/components/BaseRestController.php:
```diff
@@ -36,7 +36,7 @@ public function actionIndex()
     }
 
-    public function actionUpdate()
+    public function actionUpdate($id)
```
Should these be `$employeeId`? Since they're on the `BaseRestController`, I'm not sure.

wcjr replied:
these are just generic...non-specific to anything we've done with `employee_id`.  They're just following what would be a typical implementation, i.e., a standard view operation expects an id, not an employeeId (as constructed from the regex in the config).  Too bad Yii didn't take an annotation approach on these things, then the url matching could be built right along with the applicable classes and methods.

forevermatt replied:
In Yii1, it was typically just a standard controller/action automatic mapping, something like the following (with a few other custom routes):
```
'<controller:\w+>/<action:[\w\-]+>/<id:\d+>' => '<controller>/<action>',
'<controller:\w+>/<action:[\w\-]+>' => '<controller>/<action>',
```
In this app, there are so few endpoints (6?), spelling them out explicitly seems fine.

forevermatt replied:
Anyway, thanks for helping me understand those `$id`s here.

forevermatt commented on application/frontend/controllers/UserController.php:
```diff
@@ -4,17 +4,49 @@
+
+        $user->attributes = Yii::$app->request->getBodyParams();
+
+        parent::save($user);
```
It looks like `BaseRestController->save(...)` isn't a static method, so I think you can call it as `$this->save($user)`. You could also make it static, I suppose, which I assume would be fine.

wcjr replied:
I did it that way to be more explicit (readability) about where the call was coming from but I don't need to...chalk that up to my learning PHP as well, I've seen `::` being used in ways I didn't think meant strictly static but I can see how it might.  Shouldn't something have complained about that syntactically or at least at runtime or is this the "forgiveness" of PHP?

forevermatt replied:
Hmm... you may be right about `parent::someMethod()` not necessarily implying static. The only places I'm using to seeing `parent::` are when calling the parent version of the current method (such as calling `parent::rules()` within the current class's `rules()` method). When calling a non-static method the common practice seems to be to use `$this` regardless of whether the method is defined in the current class or a parent class.

The lack of a complaint from PHP could simply be related to special details about `parent::` (as opposed to `ClassName::`). I'm not sure on that one.

wcjr replied:
no, I think that makes sense.  Polymorphism would suggest `$this->save` is correct, I believe.  I probably did that because I was working on the model so much...but that distinction about using `parent::` while you're in an overriding method is helpful.  https://github.com/silinternational/idp-id-broker/commit/b8a62cb0a0a485b152c2fc7b6cb775cb92a1599a

forevermatt commented on application/frontend/controllers/UserController.php:
```diff
@@ -23,14 +55,14 @@ public function actionCreate(): User
 
         if ($user === null) {
-            throw new NotFoundHttpException();
+            Yii::$app->response->statusCode = 204;
```
This may be one of those situations where my push to use `204` (instead of `404`) for Not Found makes our API less intuitive. I think common practice would suggest that getting a `204` back (from a `PUT` call to update a password) would indicate success. 

Since this is following the spec we've decided on, it's probably best to leave it as `204` for now... as you said, @wcjr, we can change it later if we need to. `</second-guessing>`

wcjr replied:
no problem, at this time it feels like six of one and half-a-dozen of another but let's see if that proves out over time.  It's an easy change either way and we'll have confidence to make such a change because there will be so many tests behind it all.

wcjr commented:
resolves #32
